### PR TITLE
[Bugfix][LoRA] Avoid block_n=128 for lora_expand on Hopper

### LINKED ISSUE #48590: [Bug]: LoRA `lora_expand` Triton kernel outputs NaN on Hopper (sm_90) with `block_n=128`, producing garbled LoRA output
---

# === My current environment ===

<details>
<summary>Output of <code>python -m vllm.collect_env</code> (run inside the affected H20 container)</summary>

```text
==============================
        System Info
==============================
OS                           : Ubuntu 22.04.5 LTS (x86_64)
GCC version                  : (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Libc version                 : glibc-2.35

==============================
       PyTorch Info
==============================
PyTorch version              : 2.11.0+cu130
CUDA used to build PyTorch   : 13.0

==============================
      Python Environment
==============================
Python version               : 3.12.13
Python platform              : Linux-4.18.0-372.32.1.90.po1.x86_64-x86_64-with-glibc2.35

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : 13.0.88
GPU models and configuration : GPU 0: NVIDIA H20-3e   (Hopper, sm_90)
Nvidia driver version        : 580.105.08

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.6.12
[pip3] numpy==2.2.6
[pip3] torch==2.11.0+cu130
[pip3] torchvision==0.26.0+cu130
[pip3] transformers==5.12.0
[pip3] triton==3.6.0

vLLM Version                 : 0.23.0
```

Also reproduced on vLLM **0.24.0** and **0.25.0** — `vllm/lora/ops/triton_ops/lora_expand_op.py` and `kernel_utils.py` are byte-identical across 0.23/0.24/0.25, so upgrading does not fix it.

</details>

---

# === 🐛 Describe the bug ===

## Summary

On **NVIDIA Hopper (sm_90, e.g. H20/H100)**, serving a Qwen3-VL-8B-Instruct LoRA adapter with vLLM produces **completely garbled output** (e.g. repeated `!!!!!!`, or incoherent mixed-language/code-like tokens). The **base model without LoRA is perfectly fine**.

I traced this to the LoRA **`lora_expand`** Triton kernel producing **NaN**. `lora_shrink` is fine. The trigger is the default expand kernel config **`block_n = 128`** (single-slice path); forcing **`block_n = 32`** makes the kernel numerically correct and the garbling disappears.

It is **architecture-specific**: the exact same adapter runs correctly on **Blackwell (RTX 5070, sm_120)** and via **HF Transformers + PEFT** on the same GPU-family — only the vLLM Triton expand kernel on sm_90 is wrong.

## Transformers vs vLLM (per the issue template)

I ran the **same adapter** through plain **HF Transformers + PEFT** (`PeftModel.from_pretrained`) — output is **correct**. Only vLLM garbles. So this is a vLLM (Triton kernel) issue, not a modeling/transformers issue.

## What I ruled out (each tested, not guessed)

- vLLM version: 0.23.0 / 0.24.0 / 0.25.0 all reproduce (kernel code identical).
- `--enforce-eager` on/off: both reproduce.
- Attention backend: FlashAttn v2 / v3 / `TORCH_SDPA` all reproduce.
- Sampler backend: flashinfer sampler on/off, both reproduce.
- Quantization: bf16 / fp8 / bitsandbytes-4bit all reproduce.
- Virtualization: reproduced on **bare-metal H20** (no HAMi/vGPU).
- Stale `~/.triton/cache`: wiping it does **not** help; NaN reproduces on a fresh compile.

## Minimal, model-free reproduction

This calls only the LoRA Triton ops with synthetic tensors and compares against the torch reference (`vllm/lora/ops/torch_ops`):

```python
import torch
from vllm.lora.ops.triton_ops import lora_expand, LoRAKernelMeta
from vllm.lora.ops.torch_ops import bgmv_expand

dev = "cuda"
T, H, R, scaling = 1029, 4096, 64, 2.0   # num_tokens, hidden, rank, alpha/r
for dtype in (torch.bfloat16, torch.float16):
    x   = torch.randn(1, T, R, device=dev, dtype=dtype)          # shrink output [slices, tokens, rank]
    B   = (torch.randn(1, H, R, device=dev) * 0.02).to(dtype)    # lora_b [num_loras, hidden, rank]
    out = torch.zeros(T, H, device=dev, dtype=dtype)
    tlm = torch.zeros(T, dtype=torch.int32, device=dev)          # all tokens -> lora 0
    meta = LoRAKernelMeta.make(max_loras=1, max_num_tokens=T, device=dev)
    meta.prepare_tensors(tlm)

    lora_expand(x, [B], out, *meta.meta_args(T, False), offset_start=0, add_inputs=True)

    ref = torch.zeros(T, H, device=dev, dtype=dtype)
    bgmv_expand(x[0], B, ref, torch.zeros(T, dtype=torch.long, device=dev), add_inputs=True)

    print(dtype, "| lora_expand(triton) NaNs:", torch.isnan(out).sum().item(),
          "| bgmv_expand(torch) NaNs:", torch.isnan(ref).sum().item())
```

Observed on H20 (sm_90):

```text
torch.bfloat16 | lora_expand(triton) NaNs: 4480   | bgmv_expand(torch) NaNs: 0
torch.float16  | lora_expand(triton) NaNs: 36352  | bgmv_expand(torch) NaNs: 0
```

Only a subset of output elements are NaN (looks like a boundary N-tile), then it propagates through the forward pass → garbled decode.

## The trigger: `block_n = 128`

In `vllm/lora/ops/triton_ops/utils.py`, the default single-slice expand config is:

```python
"block_n": 64 if num_slices > 1 else 128,
```

Changing `128` → `32` yields **0 NaN** and matches the torch reference (max abs diff ~1e-3, normal bf16 rounding). So the fault is specific to the **`block_n=128` tile on sm_90** — pointing at a Triton codegen / memory-access bug at that tile size (possible out-of-bounds read on the boundary N-tile, or an MMA/shared-memory layout issue for a 128-wide tile on Hopper). Compare with #47303 (Triton MoE kernels producing NaN due to OOB reads) — similar flavor.

## Impact / fix verified

With either workaround, garbling disappears and downstream task accuracy is restored (0.63 base → ~0.90 LoRA, matching the Transformers+PEFT reference):

1. Set single-slice expand `block_n = 32` in `utils.py` (keeps native Triton, negligible perf hit, CUDA-graph safe). **Recommended.**
2. Fall back to a torch `lora_expand` on sm_90 (correct but slower; must be written CUDA-graph-safe or capture fails with `operation not permitted when stream is capturing`).

## Ask

- Is `block_n=128` for the expand kernel expected to be valid on sm_90? The NaN strongly suggests a tiling/masking/codegen bug at that config on Hopper.
- Could the default expand config avoid `block_n=128` on sm_90, or the boundary-tile masking be fixed?

Happy to provide full serve logs and the instrumentation details.

### PR DESCRIPTION
**Pull request by elinacse**
block_n=128 triggers a NaN-producing Triton codegen bug in the single-slice lora_expand kernel on sm_90, garbling LoRA output. Blackwell isn't affected, so drop to block_n=32 on Hopper only.

Fixes #48590.

Co-authored-by: Claude

### CODE REVIEW

**Comment by elinacse** [comment]
@jeejeelee Could you please review this PR ? Thanks !

**Comment by elinacse** [comment]
@jeejeelee can you review this PR . Thanks !

**Comment by elinacse** [comment]
@jeejeelee Could someone add the ready label? Thanks!
